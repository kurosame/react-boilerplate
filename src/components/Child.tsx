import {
  addCount,
  getSagaCount,
  ICounterActions,
  ICounterState
} from '@/modules/counter'
import { IStates } from '@/modules/states'
import React from 'react'
import { connect } from 'react-redux'
import { bindActionCreators, Dispatch } from 'redux'
import { Action } from 'redux-actions'
import styled from 'styled-components'

interface IProps {
  state: { counter: ICounterState }
  actions: Pick<ICounterActions, 'addCount' | 'getSagaCount'>
}

export const Child = (props: IProps) => (
  <Div>
    <div>
      <span data-test="count">{props.state.counter.count}</span>
      <button data-test="add-count" onClick={() => props.actions.addCount()}>
        ADD
      </button>
    </div>
    <div>
      <span data-test="saga-count">{props.state.counter.sagaCount}</span>
      <button
        data-test="add-saga-count"
        onClick={() => props.actions.getSagaCount()}
      >
        ADD
      </button>
      â€»redux-saga sample
    </div>
  </Div>
)

const Div = styled.div`
  color: white;
  background-color: blue;
`

export default connect(
  (states: IStates) => ({ state: { counter: states.counter } }),
  (dispatch: Dispatch) => ({
    actions: {
      addCount: bindActionCreators<Action<ICounterState>, typeof addCount>(
        addCount,
        dispatch
      ),
      getSagaCount: bindActionCreators<
        Action<ICounterState>,
        typeof getSagaCount
      >(getSagaCount, dispatch)
    }
  })
)(React.memo(Child))
