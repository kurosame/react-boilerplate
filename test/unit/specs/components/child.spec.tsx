import React from 'react'
import Child from '@/components/Child'
import { render, fireEvent, cleanup } from '@testing-library/react'
import '@testing-library/jest-dom'

afterEach(cleanup)

test('Data binding props.count to count', () => {
  const wrapper = render(<Child addCount={jest.fn()} count={123}></Child>)

  expect(wrapper.getByTestId('count')).toHaveTextContent('123')
})

test('Click the add-count will call addCount', () => {
  const addCount = jest.fn()
  const wrapper = render(<Child addCount={addCount} count={123}></Child>)

  expect(addCount.mock.calls[0]).toBeUndefined()

  fireEvent.click(wrapper.getByTestId('add-count'))

  expect(addCount).toBeCalled()
  expect(addCount.mock.calls[0]).toEqual([])
})

test('Match the snapshot', () => {
  const wrapper = render(<Child addCount={jest.fn()} count={123}></Child>)

  expect(wrapper.asFragment()).toMatchSnapshot()
})
