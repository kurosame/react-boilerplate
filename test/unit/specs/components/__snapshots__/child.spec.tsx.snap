// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match the snapshot 1`] = `"<div class=\\"sc-AxjAm ewLA-Dd\\"><span data-test=\\"count\\">123</span><button data-test=\\"add-count\\">ADD</button></div>"`;
